src/charm.py

```diff
@@ -0,0 +1,241 @@
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+        self.framework.observe(
+            self.azure_provider.on.credentials_requested, self._on_azure_credentials_requested
+        )
+
+        # actions
+        self.framework.observe(
+            self.on.get_azure_connection_info_action, self.on_get_connection_info_action
+        )
+
+    def get_missing_parameters(self) -> List[str]:
+        """Returns the missing mandatory parameters that are not stored in the config."""
+        missing_options = []
+        for config_option in AZURE_MANDATORY_OPTIONS:
+            if not self.config.get(config_option):
+                missing_options.append(config_option)
+        return missing_options
+
+    def check_and_set_status(self, set_active_if_passed=False) -> bool:
+        """Check for missing parameters and set status of the unit.
+
+        If set_active_if_passed is True, the unit status will be set to Active once
+        it is found there are no missing parameters.
+        """
+        missing_options = self.get_missing_parameters()
+        if missing_options:
+            logger.warning(f"Missing parameters: {missing_options}")
+            self.unit.status = BlockedStatus(f"Missing parameters: {missing_options}")
+            return
+        try:
+            self.decode_secret_key(self.config.get("credentials"))
+        except Exception as e:
+            logger.warning(f"Error in decoding secret: {e}")
+            self.unit.status = BlockedStatus(str(e))
+            return
+
+        if set_active_if_passed:
+            self.unit.status = ActiveStatus()
+
+    def _on_start(self, _: StartEvent) -> None:
+        """Handle the charm startup event."""
+        self.check_and_set_status()
+
+    def _on_update_status(self, event: ops.UpdateStatusEvent):
+        """Handle the update status event."""
+        self.check_and_set_status(set_active_if_passed=True)
+
+    def _on_secret_changed(self, event: ops.SecretChangedEvent):
+        """Handle the secret changed event.
+
+        When a secret is changed, it is first checked that whether this particular secret
+        is used in the charm's config. If yes, the secret is decoded and relation databag
+        is updated.
+        """
+        secret = event.secret
+
+        if not self.config.get("credentials"):
+            return
+        if self.config.get("credentials") != secret.id:
+            return
+
+        try:
+            secret_key = self.decode_secret_key(secret.id)
+        except Exception:
+            secret_key = ""
+
+        if len(self.azure_provider.relations) > 0:
+            for relation in self.azure_provider.relations:
+                self.azure_provider.update_relation_data(relation.id, {"secret-key": secret_key})
+
+        self.check_and_set_status(set_active_if_passed=True)
+
+    # def _on_secret_removed(self, event: ops.SecretRemoveEvent):
```

Comment:
One of the things that I am yet unable to do is to handle the secrets being deleted.

For eg, consider the scenario:
```
juju config object-storage-integrator credentials=mycreds

juju remove-secret mycreds
```

In such cases, I have not found a way to listen to secrets being deleted because neither `secrets-changed` or `secret-removed` events are fired in this case. The `secret-removed` event looks like something that gets fired when the observer charms no longer are observing the old revision of a secret (more details [here](https://juju.is/docs/sdk/event-secret-remove) ).

Reply:
A bug has been raised for Juju for this very scenario here: https://bugs.launchpad.net/juju/+bug/2071746. Until we have this feature in Juju though, the only workaround seems to be waiting for the next `update-status` (or some other event) to occur which would fix the charm's state from "okay, secrets have been configured" to "oops, secret no longer exists".

Reply:
In addition, I am also not able to find a way for the charm to know when secrets have been granted and revoked.

For example, consider the scenario:

```
juju add-secret somesecret secret-key=xxx
# secret:xxxxxxxxxxxxx

juju config object-storage-integrator credentials=secret:xxxxxxxxxxxx
```

In this case the charm will be in blocked state with message "Permission has not been granted for secret `secret:xxxxxxxxxx`".

However, if the user then proceeds to `juju grant-secret somesecret object-storage-integrator`, the charm is not notified about this at all. The same problem exists when the permission for the secret has been revoked using `juju revoke-secret somesecret object-storage-integrator` and the charm will not be notified the secret is no longer accessible for the charm. Currently, the charm needs to wait for the next `update-status` event for it to know of this information.